networking-calico/networking_calico/agent/dhcp_agent.py

```diff
@@ -148,6 +154,146 @@ def split_endpoint_name(name):
     return tuple([p.replace('#', '-') for p in parts])
 
 
+class MTUWatcher(object):
+
+    # -----------------------------------------------------------------
+    # Methods called from DHCP agent's main thread.
+    # -----------------------------------------------------------------
+
+    def __init__(self, agent, port_handler):
+        self.agent = agent
+        self.port_handler = port_handler
+        self.mtu_by_if_name = {}
+        self.port_id_by_if_name = {}
+        self.rx_up = re.compile('[0-9]+: ([a-z0-9-]+).* mtu ([0-9]+) .* state UP')
+        self.rx_del = re.compile('Deleted [0-9]+: ([a-z0-9-]+)')
+
+    def get_mtu(self, if_name):
+        # Note, returns None if we haven't yet seen an MTU for the given
+        # interface name.
+        return self.mtu_by_if_name.get(if_name)
+
+    def watch_port(self, id, if_name):
+        self.port_id_by_if_name[if_name] = id
+
+    def unwatch_port(self, id, if_name):
+        if if_name in self.port_id_by_if_name:
+            del self.port_id_by_if_name[if_name]
+
+    def start_up(self):
+        # Start 'ip monitor link' running on a separate thread.
+        self.ip_monitor_running = False
+        eventlet.spawn(self.process_command, ['ip', 'monitor', 'link'])
+
+        # Wait until we can be sure that 'ip monitor link' is really
+        # running, and hence that we will see all further link
+        # transitions.
+        while not self.ip_monitor_running:
+            # Create and delete a dummy interface.
+            os.system("ip link add calico-dhcp-dmy type dummy")
+            os.system("ip link del calico-dhcp-dmy type dummy")
+            time.sleep(0.5)
+
+        # Now run 'ip link', to find MTU of all pre-existing interfaces.
+        self.process_command(['ip', 'link'])
```

Comment:
This is definitely a case where I believe `pyroute2` is better suited as it should do everything in-process and speak to netlink directly.
```python3
from pyroute2 import arp
from pyroute2 import IPRoute

with IPRoute() as ipr:
  links = ipr.link("dump")
  for link in links:
    # Discard things like loopback immediately...
    if link.get("ifi_type") != arp.ARPHRD_ETHER:
      continue
    # Check that the interface is a tap: has linkinfo and type TUN
    ifla_linkinfo = link.get_attr("IFLA_LINKINFO")
    if ifla_linkinfo is None:
      continue
    if ifla_linkinfo.get_attr("IFLA_INFO_KIND") != "tun":
      continue

    ifname = link.get_attr("IFLA_IFNAME")
    print(ifname)
```

```shell
$ python3 linkdump
tape9e937e3-0e
tap8fb60e9c-de
tap102c5c63-3e
```